File: ga/glm4_result_2024_6/result/tree_glm/33_tree.c
use std::cmp::Ordering;
use std::ptr::NonNull;

type SearchTree = Option<Box<TreeNode>>;

#[derive(Debug)]
struct TreeNode {
    element: i32,
    left: SearchTree,
    right: SearchTree,
}

impl TreeNode {
    fn new(x: i32) -> Box<TreeNode> {
        Box::new(TreeNode {
            element: x,
            left: None,
            right: None,
        })
    }
}

fn make_empty(tree: &mut SearchTree) {
    if let Some(node) = tree {
        make_empty(&mut node.left);
        make_empty(&mut node.right);
        *tree = None;
    }
}

fn find(x: i32, tree: &SearchTree) -> Option<NonNull<TreeNode>> {
    tree.as_ref().and_then(|node| {
        match x.cmp(&node.element) {
            Ordering::Less => find(x, &node.left),
            Ordering::Greater => find(x, &node.right),
            Ordering::Equal => Some(NonNull::from(node)),
        }
    })
}

fn find_min(tree: &SearchTree) -> Option<NonNull<TreeNode>> {
    tree.as_ref().and_then(|node| {
        node.left.as_ref().map_or_else(
            || Some(NonNull::from(node)),
            |left| find_min(&left),
        )
    })
}

fn find_max(tree: &SearchTree) -> Option<NonNull<TreeNode>> {
    tree.as_ref().and_then(|node| {
        node.right.as_ref().map_or_else(
            || Some(NonNull::from(node)),
            |right| find_max(&right),
        )
    })
}

fn insert(x: i32, tree: &mut SearchTree) {
    match tree {
        None => *tree = Some(TreeNode::new(x)),
        Some(node) => match x.cmp(&node.element) {
            Ordering::Less => insert(x, &mut node.left),
            Ordering::Greater => insert(x, &mut node.right),
            Ordering::Equal => {},
        },
    }
}

fn delete(x: i32, tree: &mut SearchTree) {
    if let Some(mut node) = tree.take() {
        match x.cmp(&node.element) {
            Ordering::Less => {
                *tree = Some(node);
                delete(x, &mut node.left);
            },
            Ordering::Greater => {
                *tree = Some(node);
                delete(x, &mut node.right);
            },
            Ordering::Equal => {
                if let Some(mut right) = node.right.take() {
                    *tree = Some(right);
                    if let Some(mut left) = node.left.take() {
                        insert(left.element, &mut right.left);
                    }
                } else {
                    *tree = node.left;
                }
            },
        }
    }
}

fn retrieve(p: NonNull<TreeNode>) -> i32 {
    p.as_ref().element
}

fn main() {
    let mut t: SearchTree = None;
    for i in 0..50 {
        insert(i, &mut t);
    }
    for i in 0..50 {
        if let Some(p) = find(i, &t) {
            assert_eq!(retrieve(p), i);
        } else {
            panic!("Error at {}", i);
        }
    }
    for i in 0..50 {
        if i % 2 == 0 {
            delete(i, &mut t);
        }
    }
    for i in 1..50 {
        if i % 2 == 1 {
            if let Some(p) = find(i, &t) {
                assert_eq!(retrieve(p), i);
            } else {
                panic!("Error at {}", i);
            }
        } else {
            if find(i, &t).is_some() {
                panic!("Error at {}", i);
            }
        }
    }
    if let Some(min) = find_min(&t) {
        println!("Min is {}, Max is {}", retrieve(min), retrieve(find_max(&t).unwrap()));
    }
}

File: ga/gpt3_5_result_2024_6/result/tree_gpt/33_tree.rs
use std::ptr;

type ElementType = i32;

struct TreeNode {
    element: ElementType,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

type Position = Option<Box<TreeNode>>;
type SearchTree = Option<Box<TreeNode>>;

fn make_empty(t: SearchTree) -> SearchTree {
    match t {
        Some(mut tree) => {
            tree.left = make_empty(tree.left.take());
            tree.right = make_empty(tree.right.take());
            None
        }
        None => None,
    }
}

fn find(x: ElementType, t: SearchTree) -> Option<&Box<TreeNode>> {
    match t {
        Some(ref tree) if x < tree.element => find(x, tree.left.as_ref()),
        Some(ref tree) if x > tree.element => find(x, tree.right.as_ref()),
        _ => t.as_ref(),
    }
}

fn find_min(t: SearchTree) -> Option<&Box<TreeNode>> {
    match t.as_ref() {
        Some(tree) => tree.left.as_ref().map_or(t.as_ref(), |left| find_min(Some(left))),
        _ => None,
    }
}

fn find_max(t: SearchTree) -> Option<&Box<TreeNode>> {
    match t.as_ref() {
        Some(tree) => tree.right.as_ref().map_or(t.as_ref(), |right| find_max(Some(right))),
        _ => None,
    }
}

fn insert(x: ElementType, t: SearchTree) -> SearchTree {
    match t {
        Some(mut tree) => {
            if x < tree.element {
                tree.left = insert(x, tree.left.take());
            } else if x > tree.element {
                tree.right = insert(x, tree.right.take());
            }
            Some(tree)
        }
        None => Some(Box::new(TreeNode {
            element: x,
            left: None,
            right: None,
        })),
    }
}

fn delete(x: ElementType, t: SearchTree) -> SearchTree {
    match t {
        Some(mut tree) => {
            if x < tree.element {
                tree.left = delete(x, tree.left.take());
            } else if x > tree.element {
                tree.right = delete(x, tree.right.take());
            } else {
                let tmp = match tree.right {
                    Some(ref mut right) => {
                        if right.left.is_none() {
                            let tmp = tree.element;
                            tree.element = right.element;
                            Some(Box::new(TreeNode {
                                element: tmp,
                                left: None,
                                right: delete(tmp, tree.right.take()),
                            }))
                        } else {
                            let min_node = find_min(tree.right.take());
                            if let Some(node) = min_node {
                                tree.element = node.element;
                            }
                            tree.right = delete(tree.element, tree.right.take());
                            None
                        }
                    }
                    None => tree.left.take(),
                };
                return tmp;
            }
            Some(tree)
        }
        None => None,
    }
}

fn retrieve(p: &Box<TreeNode>) -> ElementType {
    p.element
}

fn main() {
    let mut t: SearchTree = None;
    let mut j = 0;

    t = make_empty(t);
    for i in 0..50 {
        j = (j + 7) % 50;
        t = insert(j, t);
    }

    for i in 0..50 {
        if let Some(p) = find(i, t) {
            if retrieve(p) != i {
                println!("Error at {}", i);
            }
        }
    }

    for i in (0..50).step_by(2) {
        t = delete(i, t);
    }

    for i in (1..50).step_by(2) {
        if let Some(p) = find(i, t) {
            if retrieve(p) != i {
                println!("Error at {}", i);
            }
        }
    }

    for i in (0..50).step_by(2) {
        if let Some(_) = find(i, t) {
            println!("Error at {}", i);
        }
    }

    if let Some(min_node) = find_min(t) {
        if let Some(max_node) = find_max(t) {
            println!("Min is {}, Max is {}", retrieve(min_node), retrieve(max_node));
        }
    }
}